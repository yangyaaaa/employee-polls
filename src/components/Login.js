import { useNavigate, useLocation } from "react-router-dom";
import { useState } from "react";
import { setAuthedUser } from "../actions/authedUser";
import { connect } from "react-redux";
import "./App.css";

const Login = (props) => {
  const navigate = useNavigate();
  const {state} = useLocation();
  const [userName, setUserName] = useState("");
  const [password, setUserPassword] = useState("");
  const [errorMessage, setErrorMessage] = useState("");


  const autheduser = Object.values(props.users).filter((user) => {
    return user.name === userName && user.password === password;
  });

  const handleLogIn = (e) => {
    e.preventDefault();
    if (!autheduser || userName === "" || password === "" || autheduser.length === 0) {
      setErrorMessage("Invalide user Log In");
    } else {
      console.log("autheduser: ", autheduser);
      props.dispatch(setAuthedUser(autheduser[0].id));
      setErrorMessage("success");
      console.log("location.pathname", state?.pathname);
      navigate(state?.pathname || "/");
    }
  };
  

  const handleUserName = (e) => {
    const user = e.target.value;
    console.log(user);
    setUserName(user);
  };

  const handleUserPassword = (e) => {
    const password = e.target.value;
    console.log(password);
    setUserPassword(password);
  };

  return (
    <div>
      <img
          src={
            "https://i.ibb.co/9pMsBJX/employee.png"
          }
          alt= 'employee poll'
          className="login-pic"
        />
        <h2>Log In</h2>
      <form onSubmit={handleLogIn}>
        <input
          data-testid="testId-name-input"
          type="text"
          className="login-form-input"
          placeholder={"Enter User Name"}
          onChange={handleUserName}
        />
        <input
          data-testid="testId-password-input"
          type="password"
          className="login-form-input"
          placeholder={"Enter Password"}
          onChange={handleUserPassword}
        />
        <input
          data-testid="testId-submit-button"
          type="submit"
          className="login-form-input"
        />
        <span
          id="errorMessage"
          style={{ color: "red" }}
          data-testid="error-notifier"
        >
          {errorMessage}
        </span>
      </form>
    </div>
  );
};

const mapStateToProps = (props) => props;

export default connect(mapStateToProps)(Login);
