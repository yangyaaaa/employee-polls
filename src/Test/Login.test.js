import { render, fireEvent, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import Login from "../components/Login";
import userReducer from "../../reducers/users";
import authedUserReducer from "../../reducers/authedUser";

describe("<Login />", () => {
  let store;
  let component;

  beforeEach(() => {
    store = configureStore({
      reducer: {
        users: userReducer,
        authedUser: authedUserReducer,
      },
    });

    store.dispatch({
      type: "users/add",
      payload: {
        id: "sarahedo",
        name: "Sarah Edo",
        password: "password123",
      },
    });

    component = render(
      <Provider store={store}>
        <Login />
      </Provider>
    );
  });

  it("should render without crashing", () => {
    expect(component).toBeTruthy();
  });

  it("should update username select value and password on change", async () => {
    const usernameSelect = component.getByTestId("testId-name-input");
    fireEvent.change(usernameSelect, { target: { value: "Sarah Edo" } });

    const passwordInput = component.getByTestId("testId-password-input");

    await waitFor(() => expect(usernameSelect.value).toBe("Sarah Edo"));

    // password field should be updated automatically when username is selected
    expect(passwordInput.value).toBe("password123");
  });

  it("should not allow the password field to be manually edited", () => {
    const passwordInput = component.getByTestId("testId-password-input");
    fireEvent.change(passwordInput, { target: { value: "newPassword" } });
    expect(passwordInput.value).toBe("password123");
  });

  it("should display error message when trying to log in without valid credentials", async () => {
    const usernameSelect = component.getByTestId("testId-name-input");
    fireEvent.change(usernameSelect, { target: { value: "Invalid User" } });

    const submitButton = component.getByTestId("testId-submit-button");
    fireEvent.click(submitButton);

    const errorMessage = await component.findByTestId("error-notifier");

    expect(errorMessage.textContent).toBe("Invalide user Log In");
  });

  it("should successfully log in with valid credentials", async () => {
    const usernameSelect = component.getByTestId("testId-name-input");
    fireEvent.change(usernameSelect, { target: { value: "Sarah Edo" } });

    const submitButton = component.getByTestId("testId-submit-button");
    fireEvent.click(submitButton);

    await waitFor(() =>
      expect(store.getState().authedUser).toBe("sarahedo")
    );
  });
});
