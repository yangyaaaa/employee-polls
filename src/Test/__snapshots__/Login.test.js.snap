// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should match previous snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <img
          alt="employee poll"
          class="login-pic"
          src="https://i.ibb.co/9pMsBJX/employee.png"
        />
        <h2>
          Log In
        </h2>
        <form>
          <select
            class="login-form-input"
            data-testid="testId-name-input"
          >
            <option
              disabled=""
              selected=""
              value=""
            >
              Select User Name
            </option>
            <option
              value="Sarah Edo"
            >
              Sarah Edo
            </option>
            <option
              value="Tyler McGinnis"
            >
              Tyler McGinnis
            </option>
            <option
              value="Mike Tsamis"
            >
              Mike Tsamis
            </option>
            <option
              value="Zenobia Oshikanlu"
            >
              Zenobia Oshikanlu
            </option>
          </select>
          <input
            class="login-form-input"
            data-testid="testId-password-input"
            placeholder="Enter Password"
            readonly=""
            type="password"
            value=""
          />
          <input
            class="login-form-input"
            data-testid="testId-submit-button"
            type="submit"
          />
          <span
            data-testid="error-notifier"
            id="errorMessage"
            style="color: red;"
          />
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <img
        alt="employee poll"
        class="login-pic"
        src="https://i.ibb.co/9pMsBJX/employee.png"
      />
      <h2>
        Log In
      </h2>
      <form>
        <select
          class="login-form-input"
          data-testid="testId-name-input"
        >
          <option
            disabled=""
            selected=""
            value=""
          >
            Select User Name
          </option>
          <option
            value="Sarah Edo"
          >
            Sarah Edo
          </option>
          <option
            value="Tyler McGinnis"
          >
            Tyler McGinnis
          </option>
          <option
            value="Mike Tsamis"
          >
            Mike Tsamis
          </option>
          <option
            value="Zenobia Oshikanlu"
          >
            Zenobia Oshikanlu
          </option>
        </select>
        <input
          class="login-form-input"
          data-testid="testId-password-input"
          placeholder="Enter Password"
          readonly=""
          type="password"
          value=""
        />
        <input
          class="login-form-input"
          data-testid="testId-submit-button"
          type="submit"
        />
        <span
          data-testid="error-notifier"
          id="errorMessage"
          style="color: red;"
        />
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
